File: phantom-mac-app/Sources/Phantom/Services/LiveCopilotRetrievePolicy.swift
import Foundation

enum LiveCopilotRetrievePolicy {
    private static let personalQuestionTypes: Set<String> = [
        "behavioral", "personal_factual", "motivation_fit", "situational"
    ]
    private static let personalEntityTypes: Set<String> = [
        "profile", "experience", "project", "document"
    ]
    private static let personalBases: Set<String> = [
        "profile_synthesis", "exact_evidence"
    ]
    private static let personalIntents: Set<String> = [
        "candidate_specific", "hybrid"
    ]

    static func shouldForceRetrieve(
        decision: LiveTurnDecision,
        retrievalAvailable: Bool,
        hasActiveEvidence: Bool
    ) -> Bool {
        _ = decision
        _ = retrievalAvailable
        _ = hasActiveEvidence
        return false
    }

    static func forceRetrieve(
        _ decision: LiveTurnDecision,
        questionText: String,
        preferredDocumentIds: [String]
    ) -> LiveTurnDecision {
        let docs = !decision.preferredDocumentIds.isEmpty
            ? decision.preferredDocumentIds
            : Array(preferredDocumentIds.prefix(8))
        return LiveTurnDecision(
            action: .retrieve,
            questionType: decision.questionType,
            intent: decision.intent,
            answerBasis: decision.answerBasis,
            entityType: decision.entityType,
            entityId: decision.entityId,
            retrievalQuery: normalizeRetrievalQuery(decision: decision, questionText: questionText),
            preferredDocumentIds: docs,
            targetSeconds: decision.targetSeconds,
            allowCode: decision.allowCode,
            confidence: decision.confidence
        )
    }

    static func normalizeRetrievalQuery(decision: LiveTurnDecision, questionText: String) -> String {
        let existing = decision.retrievalQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        if !existing.isEmpty { return String(existing.prefix(500)) }
        var parts: [String] = []
        let question = questionText
            .replacingOccurrences(of: "\n", with: " ")
            .replacingOccurrences(of: "\r", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if !question.isEmpty { parts.append(question) }
        if !decision.entityId.isEmpty { parts.append(decision.entityId) }
        let joined = parts.joined(separator: " ").trimmingCharacters(in: .whitespacesAndNewlines)
        if !joined.isEmpty { return String(joined.prefix(500)) }
        return "candidate profile experience project details"
    }

    static func queriesAreSimilar(_ left: String, _ right: String) -> Bool {
        let a = normalizeForCompare(left)
        let b = normalizeForCompare(right)
        guard !a.isEmpty, !b.isEmpty else { return false }
        if a == b { return true }
        return a.contains(b) || b.contains(a)
    }

    static func looksPersonal(_ decision: LiveTurnDecision) -> Bool {
        if personalIntents.contains(decision.intent) { return true }
        if personalEntityTypes.contains(decision.entityType) { return true }
        if personalBases.contains(decision.answerBasis) { return true }
        if personalQuestionTypes.contains(decision.questionType) {
            // Generic advice framed as situational/behavioral should stay universal.
            if decision.intent == "general",
               decision.entityType == "none",
               decision.answerBasis == "universal_knowledge" || decision.answerBasis == "universal_synthesis" {
                return false
            }
            return true
        }
        return false
    }

    private static func normalizeForCompare(_ value: String) -> String {
        value
            .replacingOccurrences(of: "\n", with: " ")
            .replacingOccurrences(of: "\r", with: " ")
            .lowercased()
            .split(whereSeparator: { $0.isWhitespace })
            .joined(separator: " ")
    }
}

File: phantom-windows-app/Services/LiveCopilotRetrievePolicy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SecureOverlay.Domain;

namespace SecureOverlay.Services
{
    public static class LiveCopilotRetrievePolicy
    {
        private static readonly HashSet<string> PersonalQuestionTypes = new(StringComparer.Ordinal)
        {
            "behavioral", "personal_factual", "motivation_fit", "situational"
        };

        private static readonly HashSet<string> PersonalEntityTypes = new(StringComparer.Ordinal)
        {
            "profile", "experience", "project", "document"
        };

        private static readonly HashSet<string> PersonalBases = new(StringComparer.Ordinal)
        {
            "profile_synthesis", "exact_evidence"
        };

        private static readonly HashSet<string> PersonalIntents = new(StringComparer.Ordinal)
        {
            "candidate_specific", "hybrid"
        };

        public static bool ShouldForceRetrieve(
            LiveTurnDecision decision,
            bool retrievalAvailable,
            bool hasActiveEvidence)
        {
            if (!retrievalAvailable || hasActiveEvidence) return false;
            if (decision.Action != LiveCopilotAction.Answer) return false;
            return LooksPersonal(decision);
        }

        public static LiveTurnDecision ForceRetrieve(
            LiveTurnDecision decision,
            string questionText,
            IReadOnlyList<string> preferredDocumentIds)
        {
            var query = BuildRetrievalQuery(decision, questionText);
            var docs = decision.PreferredDocumentIds.Count > 0
                ? decision.PreferredDocumentIds
                : (preferredDocumentIds ?? Array.Empty<string>()).Take(8).ToArray();
            return decision with
            {
                Action = LiveCopilotAction.Retrieve,
                RetrievalQuery = query,
                PreferredDocumentIds = docs
            };
        }

        public static bool LooksPersonal(LiveTurnDecision decision)
        {
            if (PersonalIntents.Contains(decision.Intent)) return true;
            if (PersonalEntityTypes.Contains(decision.EntityType)) return true;
            if (PersonalBases.Contains(decision.AnswerBasis)) return true;
            if (PersonalQuestionTypes.Contains(decision.QuestionType))
            {
                // Generic advice framed as situational/behavioral should stay universal.
                if (string.Equals(decision.Intent, "general", StringComparison.Ordinal) &&
                    string.Equals(decision.EntityType, "none", StringComparison.Ordinal) &&
                    (string.Equals(decision.AnswerBasis, "universal_knowledge", StringComparison.Ordinal) ||
                     string.Equals(decision.AnswerBasis, "universal_synthesis", StringComparison.Ordinal)))
                {
                    return false;
                }

                return true;
            }

            return false;
        }

        private static string BuildRetrievalQuery(LiveTurnDecision decision, string questionText)
        {
            if (!string.IsNullOrWhiteSpace(decision.RetrievalQuery))
                return TrimQuery(decision.RetrievalQuery);

            var parts = new List<string>();
            var question = (questionText ?? string.Empty)
                .Replace('\n', ' ')
                .Replace('\r', ' ')
                .Trim();
            if (!string.IsNullOrWhiteSpace(question)) parts.Add(question);
            if (!string.IsNullOrWhiteSpace(decision.EntityId)) parts.Add(decision.EntityId);
            var joined = string.Join(' ', parts).Trim();
            return string.IsNullOrWhiteSpace(joined)
                ? "candidate profile experience project details"
                : TrimQuery(joined);
        }

        private static string TrimQuery(string value)
        {
            var normalized = value.Trim();
            return normalized.Length <= 500 ? normalized : normalized[..500];
        }
    }
}
